### DISCUSSION TITLE
Handle audio issues like feedback (semi) automatically

### DISCUSSION THREAD

**Discussion by ann0see** [jamulussoftware/jamulus#1048] [Ideas]
Continuing #797 

In short:
How can we improve the following situation:

1. Inexperienced user connects to a Jamulus session with wrongly configured inputs/outputs —> feedback 
2. Inexperienced users haven’t set new client level
3. How do we ensure that he knows about this issue and how to fix it?

I hope that’s right?

Possible solution: automatic feedback detection/suppression?

**Comment by ann0see** [comment]
It would be difficult for a simple algorithm to detect feedback. Jamulus already indicates (latches) a clipping condition.
We could detect when a channel is "too loud" (clipping for a "long" time) and suppress it so that it doesn't make the session unusable for the other musicians. Suppress could mean turn the volume down to something very low (5%?) so that someone could troubleshoot that channel with continuous sound. Suppress could be something more drastic (e.g. kill or disconnect) but then it will be harder to troubleshoot.

@gilgongo  We can do more analysis on "abnormal" traffic if the packets are numbered. Then we can do things like (1) detect out of order packets, (2) detect erratic interpacket times, (3) detect a long delay and then see all the late packets causing xxxx, etc. This is interesting for understanding when bad Internet transport is affecting our sound. This could also be a rabbit-hole.

I think numbering the packets and collecting interarrival rates in a log could be really helpful with minimal processing. This still does not help in creating code to detect feedback. 

Is it possible some people on this thread are using the term feedback to mean some other audio defect?

_Originally posted by @gene96817 in https://github.com/corrados/jamulus/issues/797#issuecomment-752248008_

**Comment by ann0see** [comment]
Detecting feedback would be great. But Jamulus should not automatically mute or lower the volume of this user (this could become really annoying in my opinion). We could add a red warning like the "you are muted" warning which says: "Your signal has been very loud for quite some time! Do you hear feedback? Try to mute yourself and check if you‘ve set up your in/outputs correctly!" [Don‘t show this message again] [Mute myself]

**Comment by ann0see** [comment]
> Detecting feedback will take the same kind of code as eliminating feedback.
Let's clarify this thread to dealing with a user that is overwhelming the session with too much sound.
Then we can decide what conditions of "too much sound" are bad and need/deserve automatic suppression.
Some degree of automatic suppression will be needed, especially for a naive or inexperienced user that will respond slowly to muting themselves. 

> For fans of loud music, perhaps the threshold needs to be adjustable (i.e. triggering volume and duration or even disabled). Acoustic musicians probably want a lower threshold. People with tinnitus probably need a lower threshold too.

_Originally posted by @gene96817 in https://github.com/corrados/jamulus/issues/797#issuecomment-752255460_

**Comment by gene96817** [comment]
If we really want feedback detection, perhaps we should build echo cancellation into the client.
However, echo cancellation added delay. Alternatively, we could use the echo cancellation to detect echo but not insert that code in the data path. When the echo cancellation kicks in, we can use that condition to trigger the messages (or indicators) to the user. 

What is good about this approach is the client computer should have plenty of compute cycles for this. All the telephony and video conferencing apps have built-in echo cancellation. Most if this is in the codec. Does the codec we use have echo cancellation?

**Comment by ann0see** [comment]
Echo cancellation would probably result in the typical audio quality issues of video conferencing tools.

If we can use it to detect problems that‘s good. 


We should find ways to show the user what he‘s done wrong.

**Comment by gene96817** [comment]
I was referring to echo cancellation as a class of well-known algorithms that can automatically detect an echo.
The problem with video conferencing tools for singing has nothing to do with echo cancellation. It is their assumption that in a well-run meeting, only one person should be speaking at a time. Note, we do not want to run the audio to the server through the echo canceller because the echo canceller has a delay line as part of the algorithm to find the matching pattern (that is the echo). (The delay line stores the original signal for the pattern matching to find the echo.)

**Comment by ann0see** [comment]
> The delay line stores the original signal for the pattern matching to find the echo.

Thanks for the clarification!

**Comment by gilgongo** [comment]
@ann0see 

>We can do more analysis on "abnormal" traffic if the packets are numbered. 

Do we need to open a ticket for packet numbering or is that already being worked on? 

>This is interesting for understanding when bad Internet transport is affecting our sound.

By "interesting" do you mean that it would allow a user to be able to know whether (for example) to adjust their client's jitter buffer and not the server one? Or to reduce their audio quality setting rather than their buffer delay setting? Otherwise (as I keep asking), what's the point in having information if you can't reliably act on it? And more to the point - would such adjustments be able to be done automatically without user intervention? That would be cool if so...

**Comment by ann0see** [comment]
> Do we need to open a ticket for packet numbering or is that already being worked on?



No. I don't think someone is working on it. Feel free to open one

**Comment by gilgongo** [comment]
@ann0see I think an engineer needs to open that ticket really as I'm not qualified to do so.

**Comment by gene96817** [comment]
Could someone show me what the payload of the UDP packets looks like?
I'd be glad to work with someone interested in coding a solution
.

**Comment by WolfganP** [comment]
> Could someone show me what the payload of the UDP packets looks like?
> I'd be glad to work with someone interested in coding a solution

See https://github.com/softins/jamulus-wireshark/

**Comment by ann0see** [comment]
Will convert this to a discussion since I think it's not yet ready.

**Comment by JohannesBrx** [comment]
What about the following idea, which should catch the majority of audio feedback cases: After the user has connected or unmuted his microphone during a connection, we monitor the microphone input level for about 3 seconds. If the level is too high during this time or even clipping occurs, we mute the microphone and show the user a short information dialog. When the user unmutes the channel again we would not check the level again because we know that it is intentional.

In my opinion the most cases of audio feedback occur when the user connects to the server and has wrong settings. On the other hand, it is quite unlikely that the user joins a server and directly starts with full microphone volume levels. Checking the microphone level (instead of output level) has the advantage that it does not cause a false alarm when a band is jamming at a high overall output level or the mixer settings are simply wrong. The monitoring lasts for only 3 seconds after connecting/unmuting because high microphone levels might also occur during jamming and we do not want to cause false alarms.

I think the implementation of this strategy would be rather straightforward and avoid many annoying situations in Jamulus.

**Reply by gilgongo** [reply]
@JohannesBrx That sounds promising! Assuming we can get a reliable measure of "too high" during the 3 seconds, but maybe that's not an issue if the user can effectively confirm it's a false positive?

BTW on a minor implementation note: ideally we'd do this without a dialogue. Instead for example put the user's slider to zero and set it to a special (visual) state or something. The reason being that the more dialogues that you have in an application, the more they get ignored (in fact it's formally given as _"The probability of you reading a warning dialogue in a software application decreases by the inverse square of the number of other warning dialogues you see in a given usage session."_)

**Reply by ann0see** [reply]
Interesting approach especially for new and inexperienced users. 

Are you sure that muting on a high input gain should also take place if you unmute yourself after entering a session? I would only enable this feature once you have connected to a server (in unmuted state) and the input was too loud.

It might get annoying otherwise ("Why don't they hear me? I just unmuted myself?? Why am I muted again???"). I think, there should be a check box to disable this feature since it might also get annoying. We don't know what happens to streaming clients which might enter with a very loud signal?

**Reply by gilgongo** [reply]
@ann0see Aren't all those points handled by _"When the user unmutes the channel again we would not check the level again because we know that it is intentional."_? 

Also "Why don't they hear me?" would be handled by _"... put the user's slider to zero and set it to a special (visual) state"_

**Reply by ann0see** [reply]
I just edited my comment but see:

> After the user has connected **or unmuted his microphone during a connection**, we monitor the microphone input level for about 3 seconds. If the level is too high during this time or even clipping occurs, we mute the microphone

What happens if a user is muted and enters a session? I‘d suggest not to fire the loudness check since users who enter the session in muted state and unmute themselves („Mute myself button“) should be able to just manually mute themselves if they hear feedback. (I mean they know how to unmute themselves, so they should also know how to mute themselves). Maybe I misunderstood the post above?

**Reply by gilgongo** [reply]
So if the user had "mute myself" on then they would be able to hear (and hopefully) troubleshoot their feedback without bothering others when connected. So using "mute myself" would prevent auto-muting until they turned off "mute myself" or disconnected from the server?

**Reply by ann0see** [reply]
> So if the user had "mute myself" on then they would be able to hear (and hopefully) troubleshoot their feedback without bothering others when connected.

Yes. This should work. 

> So using "mute myself" would prevent auto-muting until they turned off "mute myself" or disconnected from the server?

That‘s one option. Another option would be to automatically fire mute myself and pull down the own slider to a very low volume (this would prevent feedback but might also be confusing since it stops it). We could maybe make the MUTED message on the top blinking with a message saying something like „MUTED! Your input was very loud for a long time. Do you have feedback? Check your input/output settings and then unmute yourself.[] don’t show this message again“

**Reply by gilgongo** [reply]
Yes, although "don’t show this message again" is a bit problematic. Do you mean literally never again? If the rule for auto-mute is just to trigger once until manual un-mute, seems banning it forever is slightly over the top since feedback/unexpected sound is by definition an unexpected thing that the function is helping you with :-)

**Reply by ann0see** [reply]
Yes. You‘re right. The only thing I would like to avoid would be that users can’t disable this feature. Some might not want it (streaming client,...)

**Reply by gilgongo** [reply]
OK well just a checkbox in the settings probably then -  "Disable auto-muting"?

**Reply by ann0see** [reply]
Yes. That should also be possible

**Reply by gilgongo** [reply]
@JohannesBrx @ann0see  (and anyone else) Do we think this is a viable solution? If so, we should create an Issue (or a PR) for the work, and reference this discussion.

**Reply by ann0see** [reply]
I'd wait for JohannesBrx answer. Afterwards we could open an issue.

**Reply by JohannesBrx** [reply]
I have just discovered issue #909, but have not yet read the lengthy thread completely: On my Mac, the (local) feedback is **not** suppressed when using the mute myself checkbox. I am not sure whether the behavior is the same for all platforms. This could probably replace a warning dialog since the feedback is loud enough, but is a rather inelegant for the user ;-) While the other participants could be protected, the user creating the feedback would still get the annoying feedback.

**Reply by ann0see** [reply]
> not suppressed when using the mute myself checkbox.

Yes. That's the "local monitoring" feature. 

We could turn down his local fader (if the initial bug in #909 doesn't come up) and give a clear warning where the mute myself message is.

**Reply by JohannesBrx** [reply]
I personally would prefer using the "mute myself" checkbox instead of touching the slider (when the bug would be fixed). The latter seems to me like a relatively *drastic* intervention and the user will loose the previous slider setting.

**Reply by ann0see** [reply]
> [...] The latter seems to me like a relatively drastic intervention and the user will loose the previous slider setting.

Yes. That should be good.


Problem: Parts of #909 is a feature: Local monitoring. Although another user also reported #909 I can't reproduce it anymore.

**Reply by JohannesBrx** [reply]
To prevent feedback we would have to mute the user's channel instead of activating "mute myself", right?

**Reply by ann0see** [reply]
Yes, I think so. But this would then only mute the user who has feedback in his own mix. 

Mute myself = "Send silence to the server and enable local monitoring"
Mute = (not 100% sure here) "do not mix sound of this user in my mix".

https://github.com/jamulussoftware/jamulus/discussions/1160 might be related.

**Reply by dcorson-ticino-com** [reply]
Mute Myself means the others connected don't hear the feedback.
The channel mute may or may not stop the feedback, but it will start again if the user doesn't do something about it.
From my vantage point setting Mute Myself is the right move, to protect others and motivate the user to correct the problem.

**Reply by JohannesBrx** [reply]
So one basic question is: Do we only want to protect the others from audio feedback or also the user creating the feedback?

Although a local feedback sound is a *memorable* experience for the user hopefully causing action to prevent it, I would prefer protecting both the user and the others. In my opinion this is more elegant and the better user experience. However, in that case two mute checkboxes must be activated (the *mute myself* and mute channel), which makes it more complicated for the user to recover to the previous state.

**Reply by dcorson-ticino-com** [reply]
In that case I would suggest in addition to the mute myself that we lower the users fader so that they still hear that there is a problem they need to fix, but don't cause hearing damage.

**Reply by ann0see** [reply]
Yes. That's a good idea. In this case, it has to be very clear what the user must do to recover to a "working" state.

**Reply by JohannesBrx** [reply]
> In that case I would suggest in addition to the mute myself that we lower the users fader so that they still hear that there is a problem they need to fix, but don't cause hearing damage.

When I pull down my channel slider to approx. 10%, feedback vanishes completely. In that case, it would almost have the same effect than muting the channel.

**Reply by JohannesBrx** [reply]
> @JohannesBrx @ann0see (and anyone else) Do we think this is a viable solution? If so, we should create an Issue (or a PR) for the work, and reference this discussion.

@gilgongo As soon as I have the time for it, I could prepare a draft PR such that we can test this. I would reference this discussion. Maybe a live test will give us more valuable clues about the correct strategy.

I would start off with the following basic implementation:
- detect microphone level and check for clipping (or high values) for 3 seconds after beginning a new connection

In case of feedback detected:
- activate *mute myself*
- either mute the channel or pull down the fader to a low amount
- show some kind of visual feedback to the user

Later on we may decide about whether to restart feedback detection after unmuting and all the special cases discussed here.

**Reply by ann0see** [reply]
Great thanks!

> show some kind of visual feedback to the user

I'd just use the existing red label which shows up once you're muted. Just change the text programatically (and maybe use another red tone.)

**Reply by gene96817** [reply]
Once we get this working, could we reexamine the need or a 3-second interval. Our software will hear things faster than a human. It should be possible to "hear" the condition and correct it in less than a second.

3 seconds is a long time for a new connection to disrupt a serious rehearsal in progress.

**Comment by ghost** [comment]
One would use convolution in the design an adaptive filter to suppress echo, or feedback, or background noise (some may call this deep-learning) ... Jamulus already accomplishes this without an idiot-proof method, as RTFM.
One could use AGC (automatic gain control) in the form of a compressor ... maybe.
Muting will lead to unmuting, which leads to muting, ...
